<!DOCTYPE html>
<html lang="zh">

	<head>
		<meta charset="UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<meta http-equiv="X-UA-Compatible" content="ie=edge" />
		<title>Document</title>
	</head>

	<body>
		<!--
		Promise.all方法用于将多个 Promise 实例，包装成一个新的 Promise 实例。

		var p = Promise.all([p1, p2, p3]);
		上面代码中，Promise.all方法接受一个数组作为参数，p1、p2、p3都是 Promise 实例，
		
		如果不是，就会先调用下面讲到的Promise.resolve方法，将参数转为 Promise 实例，再进一步处理。


		p的状态由p1、p2、p3决定，分成两种情况。
		
		（1）只有p1、p2、p3的状态都变成fulfilled，p的状态才会变成fulfilled，
		此时p1、p2、p3的返回值组成一个数组，传递给p的回调函数。
		
		（2）只要p1、p2、p3之中有一个被rejected，p的状态就变成rejected，
		此时第一个被reject的实例的返回值，会传递给p的回调函数。
	-->
		<script type="text/javascript">
			let p1 = new Promise((resolve,rejected)=>resolve(1)).then(value=>console.log(value));
			let p2 = new Promise((resolve,rejected)=>resolve(1)).then(value=>console.log(value));
			
			let p3 = new Promise((resolve,rejected)=>rejected('rejected'));

			Promise.all([p1, p2, p3]).then((resolve,reject) => console.log('fulfilled')).catch((value) => console.log(value))
			
			
			/*如果作为参数的 Promise 实例，自己定义了catch方法，那么它一旦被rejected，
			    并不会触发Promise.all()的catch方法。*/
		</script>
	</body>

</html>